src/back/spv/image.rs

```diff
@@ -33,36 +60,48 @@ impl<'w> BlockContext<'w> {
         coordinates: Handle<crate::Expression>,
         array_index: Option<Handle<crate::Expression>>,
         block: &mut Block,
-    ) -> Result<Word, Error> {
+    ) -> Result<TextureCoordinates, Error> {
         use crate::TypeInner as Ti;
         use crate::VectorSize as Vs;
 
         let coordinate_id = self.cached[coordinates];
+        let ty = &self.fun_info[coordinates].ty;
+        let inner_ty = ty.inner_with(&self.ir_module.types);
 
         // If there's no array index, the texture coordinates are exactly the
         // `coordinate` field of the `Expression::ImageLoad`. No work is needed.
         let array_index = match array_index {
-            None => return Ok(coordinate_id),
+            None => {
+                let value_id = coordinate_id;
+                let type_id = self.get_expression_type_id(ty);
+                let size = match *inner_ty {
+                    Ti::Scalar { .. } => None,
+                    Ti::Vector { size, .. } => Some(size),
+                    _ => return Err(Error::Validation("coordinate type")),
+                };
+                return Ok(TextureCoordinates {
+                    value_id,
+                    type_id,
+                    size,
+                });
+            }
             Some(ix) => ix,
         };
 
         // Find the component type of `coordinates`, and figure out the size the
         // combined coordinate vector will have.
-        let (component_kind, result_size) = match *self.fun_info[coordinates]
-            .ty
-            .inner_with(&self.ir_module.types)
-        {
-            Ti::Scalar { kind, width: 4 } => (kind, Vs::Bi),
+        let (component_kind, size) = match *inner_ty {
+            Ti::Scalar { kind, width: 4 } => (kind, Some(Vs::Bi)),
```

Comment:
this never returns `None` in the second part

Reply:
Yes, but it's always used in contexts that require an `Option`, and wrapping it early lets us use the structure field shorthand when we make the return value at the bottom of the function.